### PR TITLE
Introduce FileRotator, remove core dependency

### PR DESCRIPTION
**Pull request by machour**
This PR removes core dependencies from the package requirements.
It also introduce `FileRotator` with a corresponding interface for better SRP (thanks @hiqsol for the suggestion)

phpdoc still needs to be updated, but I'd like to have some feedback first.

Maybe FileRotator constructor could take less parameters & more setters.

### CODE REVIEW

**Comment by machour** [comment]
@hiqsol done, thanks for the feedback. Also added a config for `file-rotator`.

File: composer.json [hunk 1]
```diff
@@ -90,6 +89,9 @@
     "extra": {
         "branch-alias": {
             "dev-master": "3.0.x-dev"
+        },
+        "config-plugin": {
```
Comment (samdark) [inline composer.json]: Why do we need it?

File: src/FileRotator.php [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ *
+ * All existing files are moved backwards by one place, i.e., '.2' to '.3', '.1' to '.2', and so on.
+ * The property [[maxFiles]] specifies how many history files to keep.
+ */
+class FileRotator implements FileRotatorInterface
+{
+    /**
+     * @var int maximum file size, in kilo-bytes. Defaults to 10240, meaning 10MB.
+     */
+    private $maxFileSize; // in KB
```
Comment (samdark) [inline src/FileRotator.php]:
```suggestion
    private $maxFileSize;
```

File: src/FileRotator.php [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+     * around this problem.
+     */
+    private $rotateByCopy;
+
+    public function __construct(int $maxFileSize = 10240, int $maxFiles = 5, int $fileMode = null, $rotateByCopy = true)
+    {
+        $this->maxFileSize = $maxFileSize;
+        $this->maxFiles = $maxFiles;
+        $this->fileMode = $fileMode;
+        $this->rotateByCopy = $rotateByCopy && $this->isRunningOnWindows();
```
Comment (samdark) [inline src/FileRotator.php]: Wouldn't that mean that we can't rotate by copying on non-windows?

File: src/FileRotator.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    /**
+     * Sets the value of maxFileSize.
+     * @param int $maxFileSize
+     * @return FileRotator
+     */
+    public function setMaxFileSize($maxFileSize): self
```
Comment (samdark) [inline src/FileRotator.php]:
```suggestion
    public function setMaxFileSize(int $maxFileSize): self
```

File: src/FileRotator.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        return $this->maxFileSize;
+    }
+
+
+    /**
+     * Sets the value of maxFiles.
+     * @param int $maxFiles
+     */
+    public function setMaxFiles($maxFiles): self
```
Comment (samdark) [inline src/FileRotator.php]:
```suggestion
    public function setMaxFiles(int $maxFiles): self
```

File: src/FileRotator.php [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        $this->maxFiles = (int)$maxFiles;
```
Comment (samdark) [inline src/FileRotator.php]: No need for type-casting.

File: src/FileRotator.php [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+            @chmod($newFile, $this->fileMode);
+        }
+    }
+
+    /**
+     * Renames rotated file into new file
+     * @param string $rotateFile
+     * @param string $newFile
+     */
+    private function rotateByRename($rotateFile, $newFile): void
```
Comment (samdark) [inline src/FileRotator.php]:
```suggestion
    private function rotateByRename(string $rotateFile, string $newFile): void
```

File: composer.json (hunk 1 above)
Comment (machour) [reply composer.json]: So that it can be injected by DI when using Yii

**Comment by machour** [comment]
@samdark suggestions applied.

File: composer.json (hunk 1 above)
Comment (samdark) [reply composer.json]: OK.

File: src/FileRotator.php [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    private $rotateByCopy;
+
+    public function __construct(int $maxFileSize = 10240, int $maxFiles = 5, int $fileMode = null, bool $rotateByCopy = null)
+    {
+        $this->maxFileSize = $maxFileSize;
+        $this->maxFiles = $maxFiles;
+        $this->fileMode = $fileMode;
+
+        $this->rotateByCopy = $rotateByCopy;
+        if ($this->rotateByCopy === null) {
```
Comment (samdark) [inline src/FileRotator.php]: $this->rotateByCopy = $rotateByCopy?? $this->isRunningOnWindows();

**Pull request merged**